### DISCUSSION TITLE
Trying to understand update

### DISCUSSION THREAD

**Discussion by devansh08** [rebelot/heirline.nvim#213] [Q&A] [answered]
I'm trying to understand how update works.

As I'm understanding now, without any conditions/update set if I have an `init` or `provider` function it gets called on any key press (?).
But what I'm noticing is even if I set update to `false`, the same still happens. Is this not what update was supposed to prevent ?

Example:
```lua
      local ViMode = {
        init = function(self)
          self.bruh = vim.loop.gettimeofday()
        end,
        provider = function(self)
          return vim.loop.gettimeofday()
        end,
        update = false,
      }

      local SL_Left = {
        ViMode,
      }
      require("heirline").setup({
        statusline = {
          SL_Left,
        },
      })
```
Note: This example is just to try and understand how the options work.

**Comment by rebelot** [comment, marked as answer]
by default the statusline is redrawn whenever neovim thinks it is most appropriate (redraw conditions are not really documented). At each redraw fired by neovim, the entire statusline is re-evaluated. This might be inefficient for some components that need to do some heavy tasks but don't need to be re-evaluated often, maybe only when directory is changed, buffer is saved, etc. the `update` field allows to control on which events the component will be re-evaluated, instead of using its cached value. `condition`, but not `init`, is *always* evaluated, regardless. Setting `update = false` indeed is not doing what you think it should be doing, it is just saying that you don't want to specify what events should trigger the component re-evaluation, an thus is using standard internal events that trigger `redrawstatus`. If you want something to be computed at plugin initialization time, and hold a reference to it from within the components, use the `static` field.

**Reply by devansh08** [reply]
Thanks for the answer!

> `condition`, but not `init`, is _always_ evaluated, regardless.

In what cases is `init` not evaluated ?

**Reply by rebelot** [reply]
In the case that the update field is present and the condition for updating is not met.
